### ISSUE TITLE
Filesize can be reduced by putting SVG images in --css-variables

### ISSUE THREAD

**Issue by TheNewSound** [twbs/bootstrap#31703]
Some SVG's are used multiple times and the build process/minifier doesn't automatically de-duplicate identical `url("data/image ....")` being used inside CSS files.
```css
:root {
  --bs-image-form-select-indicator: escape-svg($form-select-indicator);
}

.form-select {
  background-image: var(--bs-image-form-select-indicator);
}
```

De-duping inline images by creating a css-variable out of it would save some KB's.

**Comment by mdo** [comment]
That's a very good idea! I'll look into it. Can probably do some nesting of CSS vars too given we change the full colors across a lot of these backgrounds.

**Comment by mdo** [comment]
Turns out this doesn't save as much as one would hope. Basically we go from `182985` to `182608` bytes, so we're saving ~380 bytes. This includes changing all form icons into CSS variables and using those CSS vars instead of Sass ones in the `background-image` declarations.

Output of those:

```css
:root {
  --form-select-indicator: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'><path fill='none' stroke='#343a40' stroke-linecap='round' stroke-linejoin='round' stroke-width='2' d='M2 5l6 6 6-6'/></svg>");
  --form-check-input-checked-bg-image: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 20 20'><path fill='none' stroke='#fff' stroke-linecap='round' stroke-linejoin='round' stroke-width='3' d='M6 10l3 3l6-6'/></svg>");
  --form-check-radio-checked-bg-image: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='-4 -4 8 8'><circle r='2' fill='#fff'/></svg>");
  --form-check-input-indeterminate-bg-image: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 20 20'><path fill='none' stroke='#fff' stroke-linecap='round' stroke-linejoin='round' stroke-width='3' d='M6 10h8'/></svg>");
  --form-switch-bg-image: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='-4 -4 8 8'><circle r='3' fill='rgba(0, 0, 0, 0.25)'/></svg>");
  --form-switch-focus-bg-image: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='-4 -4 8 8'><circle r='3' fill='#8bbafe'/></svg>");
  --form-switch-checked-bg-image: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='-4 -4 8 8'><circle r='3' fill='#fff'/></svg>");
  --form-feedback-icon-valid: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 8 8'><path fill='#198754' d='M2.3 6.73L.6 4.53c-.4-1.04.46-1.4 1.1-.8l1.1 1.4 3.4-3.8c.6-.63 1.6-.27 1.2.7l-4 4.6c-.43.5-.8.4-1.1.1z'/></svg>");
  --form-feedback-icon-invalid: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' width='12' height='12' fill='none' stroke='#dc3545' viewBox='0 0 12 12'><circle cx='6' cy='6' r='4.5'/><path stroke-linejoin='round' d='M5.8 3.6h.4L6 6.5z'/><circle cx='6' cy='8.2' r='.6' fill='#dc3545' stroke='none'/></svg>");
}

We'd shave a little more if we put our carousel icons in there, but limited it to one chevron and then used a CSS transform to mirror the icon for the other side.

Beyond that, not as much savings. Thoughts @twbs/css-review?

**Comment by ffoodd** [comment]
If a SVG is used only once (including it's fill) then using a custom property is counter-productive when it comes to filesize.

This should be used only for icons used at least twice, but that makes a good argument to use the same `fill` color for each icon occurence and apply colors with `filter`s as you're doing in the close icon branch.

**Comment by mdo** [comment]
Yeah, I've found that as well re: usage numbers. I think the filters are great when used surgically, but beyond that they get a little too clever maybe. Like turning black to a random color instead of just inverting/grayscaling.

Worth playing around a little more though!

**Comment by MartijnCuppens** [comment]
Custom properties in `:root` are global, which means they leak into every DOM element. I don't think there's a huge performance downside to it, but it doesn't feel like a best practice since (like @ffoodd said) they are often used only once.

`filter`s might be the way to go indeed.

**Comment by TheNewSound** [comment]
> Custom properties in `:root` are global, which means they leak into every DOM element. I don't think there's a huge performance downside to it, but it doesn't feel like a best practice since (like @ffoodd said) they are often used only once.
> 
> `filter`s might be the way to go indeed.

I don't think browsers copy the :root variables in memory for every DOM element, but I am not sure. Would sound counter-intuitive to me.

**Comment by ffoodd** [comment]
They are inherited, so yes: every single DOM element has access and can use single properties applied above in the DOM. Meaning the `:root` one are global.

**Comment by XhmikosR** [comment]
>  Basically we go from `182985` to `182608` bytes, so we're saving ~380 bytes.

To clarify, are you talking about gzip/brotli size or uncompressed?

**Comment by mdo** [comment]
Uncompressed, was just using a Terminal command to snag file size.

**Comment by XhmikosR** [comment]
We should compare the gzip output too, but I agree 100% with @ffoodd; unless we use an icon more than once, it might not benefit us a lot.

That being said, we should check the compressed output. Feel free to push a temp branch so that we check bundlewatch.

**Comment by mdo** [comment]
Closing for now as a won't fix since this doesn't seem to make a meaningful impact. Happy to revisit if other ideas come up!
